Refactor out EventBridge

mcnamarabrian opened the pull request:
*Issue #, if available:*

*Description of changes:*
This is a *big* refactor.  The deployment isn't as clean as I would've liked (see deploy.sh for a better explanation).


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bmcnamar-amazon committed:
Refactor out EventBridge

Use S3 Events and SNS in place of CloudTrail and EventBridge

jrrudge commented on deploy.sh:
```diff
@@ -0,0 +1,51 @@
+echo '' >> ${FILE_TO_UPDATE}
+sam-runner
+git checkout ${FILE_TO_UPDATE}
+echo
```
Should have a newline

jrrudge commented on template.yml:
```diff
@@ -25,112 +25,60 @@ Parameters:
             Action:
               - sqs:SendMessage
-            Resource: !GetAtt SentimentQueue.Arn
+            Resource: "*"
```
We have action: * resource * here, is that intentional?

mcnamarabrian replied:
It is.  I tried to define two separate policies and more narrowly scope the Resource but it failed (topic was unable to publish to each queue).  I used https://docs.aws.amazon.com/sns/latest/dg/SendMessageToSQS.cloudformation.html as a reference and was able to get things working.

jrrudge commented on deploy.sh:
```diff
@@ -0,0 +1,51 @@
+}
+
+###
+# Main body
```
I appreciate what you are doing here but I think it's the wrong approach, this will break really easily and it's not really usable in a pipeline. I think the input bucket is unique enough as it contains the account ID. I am not a fan of static resource names (Due to potential update issues) but in this scenario I think it is justified, and the common issues with this don't really apply to an S3 Bucket.  I have made a comment at the code that needs changing to suggest the alternative.

mcnamarabrian replied:
I'm going to remove deploy.sh entirely.  I'll leave the `cleanup.sh` script in place though.

jrrudge commented on template.yml:
```diff
@@ -25,112 +25,60 @@ Parameters:
-            Effect: Allow
-            Principal:
-              AWS: '*'
+                aws:SourceArn: !Sub "arn:aws:s3:::${InputBucket}"
```
I suggest this to 'arn:aws:s3:::${AWS::StackName}-${AWS::AccountId}-inputbucket' this is a) more in line with other naming and b) Resolves the circular dependencyf with the notification without having to use external fixes. Another approach could be a custom resource or macro backed by a lambda to generate a uuid. However I think that is overkill.

mcnamarabrian replied:
Fair enough.  I'll make the update.

jrrudge reviewed (changes requested):
It all works, I really think we should change the solution for the notification circular dependency. What you have here is probably overkill, quite fragile and not pipeline friendly.

bmcnamar-amazon committed:
Updated for better usage of Custom Resource

bmcnamar-amazon committed:
Updated to reflect build and tear-down instructions.

bmcnamar-amazon committed:
Update architecture diagram and readme with more detailed pre-reqs

bmcnamar-amazon committed:
Remove redrive policy from SQS, update ref architecture diagram

It is better to standardize on function EventInvokeConfig forerror destinations.

bmcnamar-amazon committed:
Update to reflect better ARN + change of error handling

bmcnamar-amazon committed:
Remove test messages sent from s3 to queues during stack creation

bmcnamar-amazon committed:
updatddashboard

mcnamarabrian committed:
Update boto3 clients and resources to use aws region

jrrudge commented on template.yml:
```diff
@@ -25,112 +25,89 @@ Parameters:
 Resources:
   InputBucket:
```
Would be great if we could update the input bucket name in line with the other buckets. 

  InputBucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: !Sub ${AWS::StackName}-${AWS::AccountId}-input-bucket

mcnamarabrian replied:
It is - it's updated on Line 28.

jrrudge replied:
Sorry. I should learn to read properly.

jrrudge reviewed (changes requested):
Tested and all working. Just still pending the S3 bucket rename.

jrrudge commented:
All good in that case :)

mcnamarabrian commented:
Thanks for the review, @jrrudge!
